src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Controllers/Api/Public.Data/DataSetVersionMappingControllerTests.cs

```diff
@@ -0,0 +1,530 @@
+                client);
+
+            var validationProblem = response.AssertValidationProblem();
+            Assert.Equal(3, validationProblem.Errors.Count);
+            validationProblem.AssertHasNotNullError("updates[0].level");
+            validationProblem.AssertHasNotNullError("updates[0].type");
+            validationProblem.AssertHasNotNullError("updates[0].sourceKey");
+        }
+
+        [Theory]
+        [InlineData(MappingType.ManualMapped)]
+        [InlineData(MappingType.AutoMapped)]
+        public async Task MappingTypeExpectsCandidateKey_Returns400(MappingType type)
+        {
+            var client = BuildApp().CreateClient();
+
+            var response = await ApplyBatchMappingUpdates(
+                Guid.NewGuid(),
+                [
+                    new LocationMappingUpdateRequest
+                    {
+                        Level = GeographicLevel.LocalAuthority,
+                        SourceKey = "location-1",
+                        Type = type,
+                        CandidateKey = null
+                    }
+                ],
+                client);
+
+            var validationProblem = response.AssertValidationProblem();
+            Assert.Single(validationProblem.Errors);
+            validationProblem.AssertHasNotNullError("updates[0].candidateKey");
+        }
+
+        [Theory]
+        [InlineData(MappingType.None)]
+        [InlineData(MappingType.ManualNone)]
+        [InlineData(MappingType.AutoNone)]
+        public async Task MappingTypeDoeNotExpectCandidateKey_Returns400(MappingType type)
+        {
+            var client = BuildApp().CreateClient();
+
+            var response = await ApplyBatchMappingUpdates(
+                Guid.NewGuid(),
+                [
+                    new LocationMappingUpdateRequest
+                    {
+                        Level = GeographicLevel.LocalAuthority,
+                        SourceKey = "location-1",
+                        Type = type,
+                        CandidateKey = "target-location-1"
+                    }
+                ],
+                client);
+
+            var validationProblem = response.AssertValidationProblem();
+            Assert.Single(validationProblem.Errors);
+            validationProblem.AssertHasNullError("updates[0].candidateKey");
+        }
+
+        private async Task<HttpResponseMessage> ApplyBatchMappingUpdates(
+            Guid nextDataSetVersionId,
+            List<LocationMappingUpdateRequest> updates,
+            HttpClient? client = null)
+        {
+            client ??= BuildApp().CreateClient();
+
+            var uri = new Uri($"{BaseUrl}/{nextDataSetVersionId}/mapping/locations", UriKind.Relative);
+
+            return await client.PatchAsync(uri,
+                new JsonNetContent(new BatchLocationMappingUpdatesRequest { Updates = updates }));
+        }
+    }
+
+    // permission tests (and do we have these for other controller integration tests too)?
```

Comment:
is this comment meant to be here?

Reply:
Nope!  I removed it in a subsequent PR, but I'll remove it here too :)